File: Coding-Interviews/06-从尾到头打印链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <vector>
using namespace std;
class Solution
{
public:
    vector<int> reversePrint(ListNode* head)
    {
        vector<int> ret;
        ListNode*   cur = head;
        while (cur) {
            ret.push_back(cur->val);
            cur = cur->next;
        }

        return std::move(vector<int>(ret.rbegin(), ret.rend()));
    }
};

File: Coding-Interviews/06-从尾到头打印链表.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# class Solution:
#     def reversePrint(self, head: ListNode) -> List[int]:
#         res = []
#         while head:
#             res.append(head.val)
#             head = head.next
#         return res[::-1]

# 递归写法: 先走至链表末端，回溯时依次将节点值加入列表
class Solution:
    def reversePrint(self, head: ListNode) -> List[int]:
        return self.reversePrint(head.next) + [head.val] if head else []
